docs/run_notebooks.py

```diff
@@ -0,0 +1,196 @@
+
+import matplotlib as mpl
+import matplotlib.pyplot as plt
+
+
+def execute_notebook(nb_test_dir, filename):
+    os.environ["JUPYTER_PLATFORM_DIRS"] = "1"
+
+    with open(filename) as fp:
+        nb = json.load(fp)
+
+    mpl.use("agg")  # Use a non-interactive backend
+    os.chdir(nb_test_dir)
+    notebook_state = {}
+    for cell in nb["cells"]:
+        if cell["cell_type"] == "code":
+            source = "".join(line for line in cell["source"] if not line.startswith("%"))
+            with redirect_stdout(io.StringIO()) as _, redirect_stderr(io.StringIO()) as _:
+                exec(source, notebook_state, notebook_state)
+                plt.close("all")
+
+
+def read_report(filename):
+    """Reads a dictionary from a file"""
+    if not filename.is_file():
+        return {}
+
+    try:
+        with open(filename, "r") as f:
+            return json.load(f)
+    except json.decoder.JSONDecodeError:
+        return {}
+
+
+def process_notebook(notebook, report, nb_test_dir):
+    """Process a single notebook and return a report on it
+
+    Parameters
+    ----------
+    notebook : pathlib.Path
+        Filename of a notebook
+    report : dict
+        Dictionary with notebook results
+    nb_test_dir : pathlib.Path
+        Path to the testing directory.
+    """
+    finished_success = report.get("result", "") == "success"
+
+    if finished_success:
+        print(f"Skipping notebook: {notebook} (already successful)")
+    else:
+        print(f"Testing notebook: {notebook}")
+
+        try:
+            tic = time.time()
+            execute_notebook(nb_test_dir, notebook)
+            report["time"] = f"{time.time() - tic:.2f}"
+        except Exception as e:
+            print("\nAn exception was raised:\n")
+            print(f"   {e}\n")
+            print(traceback.format_exc())
+
+            report["result"] = str(e)
+        else:
+            report["result"] = "success"
+
+    return report
+
+
+def run_notebooks(include_list, reset_cache, only_copy):
+    """Run all notebooks
+
+    Parameters
+    ----------
+    include_list : List[str]
+        Override which notebooks to run.
+    reset_cache : bool
+        Wipe the cache of which notebooks have already been run.
+    only_copy : bool
+        Only copy the notebooks to the notebook testing folder, but don't run them.
+    """
+    exclude_list = [
+        "nbwidgets",  # Exclude the notebook widgets since those require interaction
+        "cas9_kymotracking",
+    ]
+    base_dir = pathlib.Path(__file__).parent.parent.resolve()
+    nb_test_dir = base_dir / "nb_test"
+    os.makedirs(nb_test_dir, exist_ok=True)
+
+    report_file = nb_test_dir / pathlib.Path("test_report.txt")
+
+    print(f"Output folder: {nb_test_dir}")
+    notebook_folder = base_dir / "build/html/nbexport"
+
+    if not notebook_folder.exists():
+        raise FileNotFoundError(
+            "Could not find notebook folder. Are you on the root pylake folder and did you compile"
```

Comment:
```suggestion
            f"Could not find notebook folder at {str(notebook_folder)}. Are you on the root pylake folder and did you compile"
```

Reply:
Always a good idea to show the path it is looking for 👍 

Explicit `str` is not needed though since it's the default, right?

Reply:
yeah not needed, I got slightly confused due to pathlib